File: sim/damage_line.py
"""Per-line / average damage estimate from community spreadsheet model."""

from __future__ import annotations

from dataclasses import dataclass


BASE_CRIT_DAMAGE = 0.20


@dataclass(frozen=True)
class DamageInputs:
    """Ratios as fractions (0.5 = 50%), attack as flat number."""

    attack: float
    atk_percent: float = 0.0
    dmg_percent: float = 0.0
    boss_atk_percent: float = 0.0
    final_percent: float = 0.0
    skill_percent: float = 1.0  # 1.0 = 100% skill damage
    crit_rate: float = 0.0
    crit_dmg_percent: float = 0.0
    hit_count: int = 1
    is_boss: bool = True
    damage_cap: float | None = None
    crit_res: float = 0.0  # boss crit resistance (0..1)


@dataclass(frozen=True)
class DamageBreakdown:
    non_crit: float
    crit: float
    average: float
    effective_crit_rate: float


def calc_damage_line(inp: DamageInputs) -> DamageBreakdown:
    """
    Community sheet model (approx. 2020):

      NonCrit(Boss) = ATK * (1+DMG%) * (1+ATK% + Skill%*Boss%) * Skill% * (1+Final%)
      Crit          = NonCrit * (1 + 0.20 + CritDMG%)
      Average       = (1-CR)*NonCrit + CR*Crit
    """

    boss_term = inp.skill_percent * inp.boss_atk_percent if inp.is_boss else 0.0
    non_crit = (
        inp.attack
        * (1.0 + inp.dmg_percent)
        * (1.0 + inp.atk_percent + boss_term)
        * inp.skill_percent
        * (1.0 + inp.final_percent)
        * inp.hit_count
    )

    crit = non_crit * (1.0 + BASE_CRIT_DAMAGE + inp.crit_dmg_percent)

    # Crit rate vs boss crit res (soft behavior from sheet notes)
    cr = inp.crit_rate
    if inp.crit_res > 0:
        if inp.crit_res >= cr:
            effective_cr = 0.0
        else:
            effective_cr = min(cr - inp.crit_res, 0.90)  # soft cap ~90% noted for no-res bosses
    else:
        effective_cr = min(cr, 1.0)

    avg = (1.0 - effective_cr) * non_crit + effective_cr * crit

    if inp.damage_cap is not None:
        capped_crit = min(crit, inp.damage_cap)
        avg = (1.0 - effective_cr) * non_crit + effective_cr * capped_crit
        crit = capped_crit

    return DamageBreakdown(
        non_crit=non_crit,
        crit=crit,
        average=avg,
        effective_crit_rate=effective_cr,
    )

File: planner/src/engine/damageLine.ts
/** Port of sim/damage_line.py — community sheet estimate */

export const BASE_CRIT_DAMAGE = 0.2
/** Community sheet: crit line variance CD+0% … CD+50% */
export const CRIT_DMG_VARIANCE_MAX = 0.5

export interface DamageInputs {
  attack: number
  atkPercent?: number
  dmgPercent?: number
  bossAtkPercent?: number
  finalPercent?: number
  skillPercent?: number
  critRate?: number
  critDmgPercent?: number
  hitCount?: number
  isBoss?: boolean
  damageCap?: number | null
  critRes?: number
  /** Multiplier after DEF/IED (default 1 = no DEF). */
  defMultiplier?: number
}

export interface DamageBreakdown {
  nonCrit: number
  /** Mid crit (CD+0% extra) */
  crit: number
  /** Crit at CD+50% */
  critMax: number
  average: number
  effectiveCritRate: number
  minPerHit: number
  maxPerHit: number
  avgPerHit: number
}

export function calcDamageLine(inp: DamageInputs): DamageBreakdown {
  const atkPercent = inp.atkPercent ?? 0
  const dmgPercent = inp.dmgPercent ?? 0
  const bossAtkPercent = inp.bossAtkPercent ?? 0
  const finalPercent = inp.finalPercent ?? 0
  const skillPercent = inp.skillPercent ?? 1
  const critRate = inp.critRate ?? 0
  const critDmgPercent = inp.critDmgPercent ?? 0
  const hitCount = Math.max(inp.hitCount ?? 1, 1)
  const isBoss = inp.isBoss ?? true
  const critRes = inp.critRes ?? 0
  const defMult = inp.defMultiplier ?? 1

  const bossTerm = isBoss ? skillPercent * bossAtkPercent : 0
  let nonCrit =
    inp.attack *
    (1 + dmgPercent) *
    (1 + atkPercent + bossTerm) *
    skillPercent *
    (1 + finalPercent) *
    hitCount *
    defMult

  let crit = nonCrit * (1 + BASE_CRIT_DAMAGE + critDmgPercent)
  let critMax =
    nonCrit * (1 + BASE_CRIT_DAMAGE + critDmgPercent + CRIT_DMG_VARIANCE_MAX)

  let effectiveCritRate: number
  if (critRes > 0) {
    if (critRes >= critRate) effectiveCritRate = 0
    else effectiveCritRate = Math.min(critRate - critRes, 0.9)
  } else {
    effectiveCritRate = Math.min(critRate, 1)
  }

  let average =
    (1 - effectiveCritRate) * nonCrit + effectiveCritRate * crit

  if (inp.damageCap != null && inp.damageCap > 0) {
    const cappedCrit = Math.min(crit, inp.damageCap)
    const cappedMax = Math.min(critMax, inp.damageCap)
    average =
      (1 - effectiveCritRate) * nonCrit + effectiveCritRate * cappedCrit
    crit = cappedCrit
    critMax = cappedMax
  }

  return {
    nonCrit,
    crit,
    critMax,
    average,
    effectiveCritRate,
    minPerHit: nonCrit / hitCount,
    maxPerHit: critMax / hitCount,
    avgPerHit: average / hitCount,
  }
}
